### PR TITLE
[SPARK-43866] Partition filter condition should pushed down to metastore query if it is equivalence Predicate

### PR DESCRIPTION
**Pull request by zzzzming95**
### What changes were proposed in this pull request?

Typically, hive partition fields are created as string types.

```
CREATE TABLE if not exists test_tb (
id int
)
PARTITIONED BY (dt string)
```
However, cast data conversions are often introduced inadvertently during use. For example

```
select * from test_tb where dt=20230505;
```
it will prevent the condition `dt=20230505` from being pushed down into the metastore , because `20230505` is an IntegralType. And resulting in a request for all partitions from metastore.  This can affect execution performance in cases where the data table has very many partitions.

So, in the case of equivalent predicates, partition field pushdown should be supported.

### Why are the changes needed?

As described above

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Manual testing

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 1]
```diff
@@ -1041,6 +1053,9 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
           ExtractableLiteral(value), ExtractAttribute(SupportedAttribute(name))) =>
         Some(s"$value ${op.symbol} $name")
 
+      case EqualTo(ExtractStringAttribute(SupportedAttribute(name)), ExtractableLiteral(value)) =>
+        Some(s"$name = $value")
+
```
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Could you add a test to verify it?

[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: Please add a end to end test.

[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: yes.

[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: A test like this:
```scala
  test("SPARK-43866:xxx") {
    withTable("test_tb") {
      sql("CREATE TABLE test_tb (id int) PARTITIONED BY (dt string)")
      sql("insert into test_tb partition(dt = '20230505.0') select 1")
      val df = sql("select * from test_tb where dt=20230505")
      checkAnswer(df, Row(1, "20230505.0"))
    }
  }
```

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala [hunk 2]
```diff
@@ -994,6 +994,18 @@ private[client] class Shim_v0_13 extends Shim_v0_12 {
       }
     }
 
+    object ExtractStringAttribute {
+      @scala.annotation.tailrec
+      def unapply(expr: Expression): Option[Attribute] = {
+        expr match {
+          case attr: Attribute => Some(attr)
+          case Cast(child, dt: IntegralType, _, _)
+            if child.dataType.isInstanceOf[StringType] => unapply(child)
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala:1003]: This assumes Spark and Hive have the same behavior for implicit cast. Is it always true? For example, cast string to int may fail in Spark with ansi mode on, does hive have the same behavior?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveShim.scala]: This is incorrect, I told you before. I recommend that you change the type of partition column to int/date type.

**Pull request closed without merging**